### DISCUSSION TITLE
Wrongful oscillation detection

### DISCUSSION THREAD

**Discussion by gbizzotto** [hneemann/Digital#1052] [Ideas]
I've created a clockless CPU but can't run it because Digital interrupts it claiming it is oscillating.
By nature, clockless CPUs don't have changing inputs like a clock. All there is is an ON button, which propagates signals until the CPU detects a stop condition, like an unknown instruction.
I thought the oscillation detection was based on comparison of the current state to previous states. Instead, I suspect Digital has a set maximum number of steps it will perform before giving up, and input changes resets the counter.
Oscillation could be detected by concatenating the state of every component at every step, and storing a hash of that, then comparing it to previous hashes. A repeated hash means the circuit is in an infinite loop. Execution would probably be significantly slower, but it could be optional.

**Comment by hneemann** [comment]
There is support for asynchronous automata.
See `examples/sequential/async/muller-pipeline.dig` as an example.
